lib/router/src/cache.rs

```diff
@@ -1,426 +1,367 @@
-            let now = self.now.now();
-            self.vals.retain(|_, n| {
-                let age = now - n.last_access();
-                age.as_secs() <= max_age
-            });
-
-            if self.vals.len() == self.capacity {
-                return Err(CapacityExhausted {
-                    capacity: self.capacity,
-                });
+            match self.expirations.poll() {
+                // The cache is at capacity but we are able to remove a value.
+                Ok(Async::Ready(Some(entry))) => {
+                    self.vals.remove(entry.get_ref());
+                }
+
+                // `Ready(None)` can only be returned when expirations is
+                // empty. We know `expirations` is not empty because `vals` is
+                // not empty.
+                Ok(Async::Ready(None)) => unreachable!(),
+
+                // The cache is at capacity and no values can be removed.
+                Ok(Async::NotReady) => {
+                    return Err(CapacityExhausted {
+                        capacity: self.capacity,
+                    });
+                }
+
+                Err(_e) => {
+                    return Err(CapacityExhausted {
+                        capacity: self.capacity,
+                    });
+                }
             }
         }
 
         Ok(Reserve {
+            expirations: &mut self.expirations,
+            expires: self.expires,
             vals: &mut self.vals,
-            now: &self.now,
         })
     }
 
-    /// Overrides the time source for tests.
-    #[cfg(test)]
-    fn with_clock<M: Now>(self, now: M) -> Cache<K, V, M> {
-        Cache {
-            now,
-            vals: self.vals,
-            capacity: self.capacity,
-            max_idle_age: self.max_idle_age,
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(entry) = try_ready!(self.expirations.poll()) {
+            self.vals.remove(entry.get_ref());
         }
+
+        Ok(Async::Ready(()))
     }
 }
 
-// ===== impl Reserve =====
+// ===== impl PurgeCache =====
 
-impl<'a, K: Hash + Eq + 'a, V: 'a, N: Now + 'a> Reserve<'a, K, V, N> {
-    /// Stores a route in the cache.
-    pub fn store(self, key: K, val: V) {
-        let node = Node::new(val.into(), self.now.now());
-        self.vals.insert(key, node);
-    }
-}
+impl<K: Clone + Eq + Hash, V> Future for PurgeCache<K, V> {
+    type Item = ();
+    type Error = ();
 
-// ===== impl Access =====
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        try_ready!(self.interval.poll().map_err(|e| {
+            println!("error polling purge interval: {:?}", e);
+            ()
+        }));
 
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
-    }
-}
+        let lock = match self.cache.upgrade() {
+            Some(lock) => lock,
+            None => return Ok(Async::Ready(())),
```

Comment:
A comment here stating that failing to upgrade the weak ref means that the cache is dropped, so the purging cache should end might be good?